src/rec_mutex.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2019, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib_sys;
+use std::mem;
+use translate::{from_glib, Stash, ToGlibPtr};
+
+#[derive(Debug)]
+pub enum RecMutex<'a> {
+    Owned(glib_sys::GRecMutex),
+    Borrowed(&'a glib_sys::GRecMutex),
+}
+
+impl<'a> RecMutex<'a> {
+    pub fn new() -> Self {
+        let rec_mutex = unsafe {
+            let mut mutex = mem::zeroed();
+            glib_sys::g_rec_mutex_init(&mut mutex);
+            mutex
+        };
+        RecMutex::Owned(rec_mutex)
+    }
+
+    #[doc(hidden)]
+    pub fn from_glib_none(rec_mutex: &'a glib_sys::GRecMutex) -> Self {
```

Comment:
This is not really a `from_glib_none()`. It does not give us an owned version of the thing, like all other `from_glib_none` do.

Maybe instead let's call it just `borrow`? Also unclear if this shouldn't be `unsafe`. If you initialize ` GRecMutex` with random data then this will probably cause crashes, and you can safely initialize a `GRecMutex` struct with random data.

In `gstreamer-rs` I have such things as `unsafe fn from_ptr<'a>(ptr: *const X) -> &'a Y`.